### PR TITLE
[3rdparty, document] Updated Documentation that for triton fused_moe kernel tuning for AMD Instinct GPUs

### PR DESCRIPTION
**Pull request by kkHuang-amd**
## Motivation

Updated Documentation for triton fused_moe kernel tuning for AMD Instinct GPUs.

## Modifications

- Upload a tuning script file
- introduce the tuning parameters setting
- Provided example bash commands for tuning script run

## Checklist

- [O] Update documentation as needed, including docstrings or example tutorials.

### CODE REVIEW

File: 3rdparty/amd/tuning/benchmark_moe_rocm.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        prog="benchmark_mixtral_moe",
+        description="Benchmark and tune the fused_moe kernel",
+    )
+    parser.add_argument(
+        "--dtype",
+        type=str,
+        default="auto",
+        choices=["float8", "float16"],
```
Comment (HaiShaw) [inline 3rdparty/amd/tuning/benchmark_moe_rocm.py]: and `bfloat16`

File: 3rdparty/amd/tuning/TUNING.md [hunk 2]
```diff
@@ -92,6 +92,20 @@ TORCHINDUCTOR_MAX_AUTOTUNE=1 TORCHINDUCTOR_COORDINATE_DESCENT_TUNING=1 TORCHINDU
...
+### Key parameters:
+- **--model**: what moe model type to do tuning, it will automatically decide the size of d_model, model_intermediate_size, num_layers
+- **--tp-size**: simulate the whole model run configuration to set the dimension size using tp correctly
+- **--batch**: M dimension size of moe kernel, for prefill moe kernel the value is batch*input_len, for decode moe kernel the value is batch
+- **--dtype**: computation type
+
+```bash
+#Tuning
+python benchmark_moe_rocm.py --model grok1 --tp-size 8 --dtype float8 --batch "32"
+
```
Comment (HaiShaw) [inline 3rdparty/amd/tuning/TUNING.md]: Provide concrete examples to tune prefill and/or decode kernels from what `--batch` is defined?

File: 3rdparty/amd/tuning/benchmark_moe_rocm.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+            int(bs), model=model, method=method, gpu=gpu, tp_size=tp_size, dtype=dtype
+        )
+
+
+def prune_configs(M, N, K, configs):
+    pruned_configs = []
+    elemBytes_a = 1  # [DV Note] Hard-coded for float16 (2 bytes)
+    elemBytes_b = 1  # [DV Note] Hard-coded for float16 (2 bytes)
+
+    mfma = 16 if M < 32 or N < 32 else 32
```
Comment (HaiShaw) [inline 3rdparty/amd/tuning/benchmark_moe_rocm.py:30]: `4x4` when it is supported?

File: 3rdparty/amd/tuning/benchmark_moe_rocm.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    method,
+    config,
+    dtype: str,
+) -> float:
+    shard_intermediate_size = model_intermediate_size // tp_size
+
+    hidden_states = torch.rand(
+        (bs, d_model),
+        device="cuda:0",
+        dtype=torch.float16,
```
Comment (HaiShaw) [inline 3rdparty/amd/tuning/benchmark_moe_rocm.py]: `hidden_states` got to use `torch_dtype` as in model config, in grok1: `bfloat16`

File: 3rdparty/amd/tuning/benchmark_moe_rocm.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+        hidden_states = method(
+            hidden_states=hidden_states,
+            w1=w1,
+            w2=w2,
+            w1_scale=w1_scale,
+            w2_scale=w2_scale,
+            a1_scale=a1_scale,
+            a2_scale=a2_scale,
+            gating_output=gating_output[0],
+            topk=2,
```
Comment (HaiShaw) [inline 3rdparty/amd/tuning/benchmark_moe_rocm.py]: use `top_k` instead of `2`

File: 3rdparty/amd/tuning/benchmark_moe_rocm.py [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+        d_model = 6144
+        model_intermediate_size = 16384
+        num_layers = 56
+    elif model == "grok1":
+        d_model = 6144
+        model_intermediate_size = 32768
+        num_layers = 64
+    else:
+        raise ValueError(f"Unsupported Mixtral model {model}")
+    num_total_experts = 8
```
Comment (HaiShaw) [inline 3rdparty/amd/tuning/benchmark_moe_rocm.py]: suggest to move this before `model` branches, and we could reassign it in specific model, when `num_total_experts != 8`, so as `top_k`.

**Review by HaiShaw — CHANGES_REQUESTED** [review]

File: 3rdparty/amd/tuning/benchmark_moe_rocm.py (hunk 3 above)
Comment (kkHuang-amd) [reply 3rdparty/amd/tuning/benchmark_moe_rocm.py:30]: The Triton team does not give the specific date for 4x4 support

File: 3rdparty/amd/tuning/benchmark_moe_rocm.py [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+import json
+import os
+import sys
+
+import torch
+import torch.nn.functional as F
+import triton
+import triton.language as tl
+from tqdm import tqdm
+from vllm import _custom_ops as ops
```
Comment (ispobock) [inline 3rdparty/amd/tuning/benchmark_moe_rocm.py]: I noticed that this custom ops dependency is not used. Could we remove it?

File: 3rdparty/amd/tuning/benchmark_moe_rocm.py [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+
+from sglang.srt.layers.fused_moe.fused_moe import fused_moe, get_config_file_name
```
Comment (ispobock) [inline 3rdparty/amd/tuning/benchmark_moe_rocm.py]: The path for this fused_moe module has changed due to recent reorganization. We need to update the import path.

File: 3rdparty/amd/tuning/benchmark_moe_rocm.py [hunk 9]
```diff
@@ -0,0 +1,380 @@
...
+    temp_list.extend(l2)
+    for myDict in temp_list:
+        if myDict not in result:
+            result.append(myDict)
+
+    return result
+
+
+def run_grid(bs, model, method, gpu, tp_size, dtype: str):
+    if model == "8x7B":
```
Comment (ispobock) [inline 3rdparty/amd/tuning/benchmark_moe_rocm.py]: Perhaps we can read these settings from model config to make it more general. [ref](https://github.com/vllm-project/vllm/blob/1f6584ee851501cfae672973b9e55d000729818c/benchmarks/kernels/benchmark_moe.py#L278)

**Comment by ispobock** [comment]
Thanks for contributing this great tuning script! Could we use this script to tune on other devices? Any general steps for adaptation?

**Comment by kkHuang-amd** [comment]
> Thanks for contributing this great tuning script! Could we use this script to tune on other devices? Any general steps for adaptation?

Yes, it can be used on other devices, but some parameter is the ROCm platform only, it needs to be removed when using on other platforms.

These two parameters only exist on ROCm
1) matrix_instr_nonkdim_range
2) kpack_range

**Comment by HaiShaw** [comment]
@ispobock Later we may come up with a generic script for more devices, after we collect device specific triton kargs extensions.
If this one merged first, we can modify the script pointing to the newer one later.

**Review by ispobock — APPROVED** [review]

File: 3rdparty/amd/tuning/benchmark_moe_rocm.py [hunk 10]
```diff
@@ -0,0 +1,377 @@
...
+import sys
+
+import torch
+import torch.nn.functional as F
+import triton
+import triton.language as tl
+from tqdm import tqdm
+from transformers import AutoConfig
+
+from sglang.srt.layers.fused_moe_grok.fused_moe import fused_moe, get_config_file_name
```
Comment (zhyncs) [inline 3rdparty/amd/tuning/benchmark_moe_rocm.py:13]: hold on
`fused_moe_grok` will be removed soon

Comment (zhyncs) [reply 3rdparty/amd/tuning/benchmark_moe_rocm.py:13]: ref https://github.com/sgl-project/sglang/pull/2223

**Comment by zhyncs** [comment]
ref https://github.com/sgl-project/sglang/pull/2225

**Review by HaiShaw — APPROVED** [review]
Thanks!

**Comment by zhyncs** [comment]
Thanks @kkHuang-amd @HaiShaw May you help update the doc from `fused_moe_grok` to `fused_moe_triton` in follow-up PR? Thanks!

**Comment by HaiShaw** [comment]
@zhyncs certainly, also notified @merrymercy for #2223

**Pull request merged**